framework/src/modules/chain/logic/broadcaster.js

```diff
@@ -95,135 +92,97 @@ class Broadcaster {
 	 * Calls peers.list function to get peers.
 	 *
 	 * @param {Object} params
-	 * @param {function} cb
 	 * @returns {SetImmediate} error, peers
 	 * @todo Add description for the params
 	 */
-	getPeers(params, cb) {
+	getPeers(params) {
 		params.limit = params.limit || this.config.peerLimit;
 		const peers = library.logic.peers.listRandomConnected(params);
-		return setImmediate(cb, null, peers);
+		return peers;
 	}
 
 	/**
 	 * Gets peers and for each peer create it and broadcast.
 	 *
 	 * @param {Object} params
 	 * @param {Object} options
-	 * @param {function} cb
-	 * @returns {SetImmediate} error, peers
+	 * @returns {Promise}
+	 * @throws {Error}
 	 * @todo Add description for the params
 	 */
-	broadcast(params, options, cb) {
+	async broadcast(params, options) {
+		let peers;
 		params.limit = params.limit || this.config.broadcastLimit;
-
-		async.waterfall(
-			[
-				function getPeers(waterCb) {
-					if (!params.peers) {
-						return self.getPeers(params, waterCb);
-					}
-					return setImmediate(
-						waterCb,
-						null,
-						params.peers.slice(0, params.limit)
-					);
-				},
-				function sendToPeer(peers, waterCb) {
-					library.logger.debug('Begin broadcast', options);
-					peers.forEach(peer => peer.rpc[options.api](options.data));
-					library.logger.debug('End broadcast');
-					return setImmediate(waterCb, null, peers);
-				},
-			],
-			(err, peers) => {
-				if (cb) {
-					return setImmediate(cb, err, { peers });
-				}
-				return true;
+		try {
+			if (!params.peers) {
+				peers = this.getPeers(params);
+			} else {
+				peers = params.peers.slice(0, params.limit);
 			}
-		);
-	}
-}
 
-// TODO: The below functions should be converted into static functions,
-// however, this will lead to incompatibility with modules and tests implementation.
-/**
- * Adds new object {params, options} to queue array.
- *
- * @param {Object} params
- * @param {Object} options
- * @returns {Object[]} Queue private variable with new data
- * @todo Add description for the params
- */
-Broadcaster.prototype.enqueue = function(params, options) {
-	options.immediate = false;
-	return self.queue.push({ params, options });
-};
+			// Broadcast using Elements P2P library via network module
+			await this.channel.invoke('network:send', {
```

Comment:
I was expecting this and I don't like this approach. I already shared many times that avoid using `actions` and try more `events`. With these action approach the `chain` module is dependent on the `network` module. Although the original idea of modules were they must be pluggable. Chain module works as it is, we plugin network module and single node now have connection to others. 

So first I suggest the approach to remove all `peers` information from the chain module and then afterwards try more `events`. e.g. What you want to broadcast a *new block* create an event for it. Then use `channel.publish('chain:block:new', block);`, on network module listen to this event and then do what ever you want with it, publish or discard or what ever you want. 

If required we should discuss in detail but ty our self to be more align with *event driven* approach. That's not just for the this module, this practice should be followed fo all modules. @MaciejBaj @shuse2

Reply:
I agree to the even driven approach, butt this is specific case where the chain module needs to tell exactly what needs to be done.
Network module is not listening to any chain specific event now.

Reply:
In this case, I think that using `invoke` is better for sending data to the network because it allows the Network module to remain module-agnostic (ideally, the Network module should not need to know ahead of time what modules or custom events are available on the node).

That way node operators will be able to add any custom modules to their nodes and those modules will be able to send custom requests and messages to each other directly over the network.

The current principle is that modules which need to send messages over the network need to be aware of the Network module but not the other way around. This dependency makes sense IMO because you can't have a Dapp without some kind of network layer.

If testability is the concern (e.g. ability to run a node without the network for testing purposes), that should not be a problem; it's still possible mock the Network module for testing if necessary.

Also, the Network module can still be substituted for other kinds of network modules (e.g. with different transports, P2P protocols, network topology, etc...) provided that they adhere to the same simple interface. The module currently only exposes about 8 different methods and this interface should serve as a standard for any third-party who wants to create a custom network module.

Reply:
> Network module to remain module-agnostic 

That also applies to other modules as well, network module is no exception. 

> Network module can still be substituted for other kinds

Yes it can be substituted even in the event driven approach. 

> The current principle is that modules which need to send messages over the network need to be aware of the Network

That's where the basic different of understanding is. Network module itself should encapsulate all network related logic, other modules do the job of their concerns. I am saying this from very start, network module is not a network proxy. It should and must encapsulate all business logic related to network communication of node. Should listen to activities in system and perform network operations, and serve network related information to other modules. 

If you want to use it just as proxy then why not just drop in the library to every module and use network functionality as library.

Reply:
If we make the Network module consume events published by CustomModule, the Network module will depend on CustomModule. It's a loose dependency because we could later replace CustomModule with NewCustomModule (if NewCustomModule publishes the same events in the same format).

However, the dependency problem in this scenario is not about whether CustomModule can be replaced with another module (it can), the problem is that the Network module needs to know about the existence of CustomModule to begin with. The Network module should not know which modules exist on the node; that way any new custom third-party module can be added later without having to change the Network module at all.

> If you want to use it just as proxy then why not just drop in the library to every module and use network functionality as library.

- It would create a lot of duplicate connections (since each module would maintain its own separate network with its own peer connections).
- There needs to be a way to ban malicious peers from a central place on the node. Many modules can report violations but the final decision to ban a peer (and for how long) must be done in one place.
- There needs to be a single consistent P2P protocol for all modules to share. If each module ran a potentially different version of the `lisk-p2p` library, we would be forcing module authors to manage their own P2P network versioning. One benefit of having a single network protocol shared by many modules is interoperability between different modules at the network level. We can upgrade the P2P protocol (and make significant changes) without changing the Network module API; so any third-party custom module created by community members do not need to change when we upgrade the P2P protocol.

Reply:
Module concept is clear, it must have its own special domain. To me proxying stuff around does not fit in the definition of a module. We will refactor this concept as separate issue https://github.com/LiskHQ/lisk-sdk/issues/3358